eZ/Publish/Core/Persistence/Legacy/Content/Type/ContentUpdater/Action/AddField.php

```diff
@@ -77,43 +77,66 @@ public function __construct(
     {
-        $languageCodeSet = array();
-        $versionNumbers = $this->contentGateway->listVersionNumbers( $contentInfo->id );
+        $versionNumbers = $this->contentGateway->listVersionNumbers( $contentId );
+        $languageCodeToFieldId = array();
 
-        $contentRows = $this->contentGateway->load( $contentInfo->id, $contentInfo->currentVersionNo );
-        $contentList = $this->contentMapper->extractContentFromRows( $contentRows );
-        $content = $contentList[0];
-
-        foreach ( $content->fields as $field )
+        foreach ( $versionNumbers as $versionNo )
         {
-            if ( isset( $languageCodeSet[$field->languageCode] ) )
-            {
-                continue;
-            }
+            $contentRows = $this->contentGateway->load( $contentId, $versionNo );
+            $contentList = $this->contentMapper->extractContentFromRows( $contentRows );
+            $content = $contentList[0];
+            $languageCodeSet = array();
 
-            $languageCodeSet[$field->languageCode] = true;
-
-            foreach ( $versionNumbers as $versionNo )
+            // Each subsequent Content version can have additional language(s)
+            foreach ( $content->fields as $field )
             {
-                $this->insertField(
+                $languageCode = $field->languageCode;
+
+                // Add once for each language per version
+                if ( isset( $languageCodeSet[$languageCode] ) )
+                {
+                    continue;
+                }
+
+                $languageCodeSet[$languageCode] = true;
+
+                // Check if field was already inserted for current language code,
+                // in that case we need to preserve its ID across versions
+                if ( isset( $languageCodeToFieldId[$languageCode] ) )
+                {
+                    $fieldId = $languageCodeToFieldId[$languageCode];
+                }
+                else
+                {
+                    $fieldId = null;
+                }
+
+                $languageCodeToFieldId[$languageCode] = $this->insertField(
                     $content,
-                    $this->createField( $versionNo, $field->languageCode )
+                    $this->createField(
+                        $fieldId,
+                        $versionNo,
+                        $languageCode
+                    )
                 );
             }
         }
     }
 
     /**
-     * Inserts given $field and appends it to the given $content field collection.
+     * Inserts given $field to the internal and external storage.
+     *
+     * If $field->id is null, creating new field id will be created.
+     * Otherwise it will be inserted for the given $content version, reusing existing Field id.
      *
      * @param \eZ\Publish\SPI\Persistence\Content $content
      * @param \eZ\Publish\SPI\Persistence\Content\Field $field
      *
-     * @return void
+     * @return int $fieldId
```

Comment:
Thanks, fixed in https://github.com/ezsystems/ezpublish-kernel/commit/5c46c9ffc9699644da7ef44896e9f9f85b6c728d.